contributors/design-proposals/storage/local-volume-management.md

```diff
@@ -0,0 +1,219 @@
+
+## <a name="roadmap"></a>Roadmap
+
+* Volume relative interface refactor
+	* Refactor volume plugin interfaces(`AttachableVolumePlugin`) and volume interfaces(`Attacher/Detacher`)
+	* Refactor `A/D controller` and `Kubelet VolumeManager` to use new interfaces
+* Refactor local volumes to implement `DevicemountableVolumePlugin` only
+	* Refactor `rbd`
+	* Refactor `fc`, `iscsi`, etc.
+
+## <a name="design"></a>Implementation Design
+
+### <a name="k8s_refactor"></a>Kubernetes Side Refactor
+
+#### <a name="k8s_interfaces_refactor"></a>Volume Plugin Interfaces refactor
+
+We extract the device mount relative methods from `Attacher` to a new interface, and embed this new interface back to `Attacher` so that volumes implement `Attacher` need also implement `DeviceMounter`:
+
+```go
+// Attacher can attach a volume to a node.
+type Attacher interface {
+	DeviceMounter
+
+	// Attaches the volume specified by the given spec to the node with the given Name.
+	// On success, returns the device path where the device was attached on the
+	// node.
+	Attach(spec *Spec, nodeName types.NodeName) (string, error)
+
+	// VolumesAreAttached checks whether the list of volumes still attached to the specified
+	// node. It returns a map which maps from the volume spec to the checking result.
+	// If an error is occurred during checking, the error will be returned
+	VolumesAreAttached(specs []*Spec, nodeName types.NodeName) (map[*Spec]bool, error)
+
+	// WaitForAttach blocks until the device is attached to this
+	// node. If it successfully attaches, the path to the device
+	// is returned. Otherwise, if the device does not attach after
+	// the given timeout period, an error will be returned.
+	WaitForAttach(spec *Spec, devicePath string, pod *v1.Pod, timeout time.Duration) (string, error)
+}
+
+// DeviceMounter can mount a volume to node.
+type DeviceMounter interface {
+	// GetDeviceMountPath returns a path where the device should
+	// be mounted after it is attached. This is a global mount
+	// point which should be bind mounted for individual volumes.
+	GetDeviceMountPath(spec *Spec) (string, error)
+
+	// MountDevice mounts the disk to a global path which individual pods
+	// can then bind mount. Returns the global mount path if succeeded,
+	// otherwise returns an error.
+	MountDevice(spec *Spec, devicePath string, pod *v1.Pod) (string, error)
+}
+```
+
+A `*v1.Pod` pararmeter is added for `MountDevice` method because we need to use pod's namespace to fetch according secret if the volume is a pod inline volume, or the secret namespace is not set.
```

Comment:
Why it needs the Secrets at this point? If it's for Ceph RBD, the volume must be attached in `WaitForAttach()`, because after that it could be used as a raw block volume and MountDevice won't be called for it.

Perhaps we could rename `WaitForAttach()` to `CompleteNodeAttach()`, that would do local attach (if necessary) and find the block device (for both local and remote attach).